lib/ecto/changeset/relation.ex

```diff
@@ -244,7 +248,7 @@ defmodule Ecto.Changeset.Relation do
       single_change(new, current, fun, [:update, :delete], true)
     else
       case on_replace(relation, current) do
-        {:ok, _} -> single_change(new, nil, fun, [:insert], false)
+        {:ok, changeset} -> single_change(new, nil, fun, [changeset.action], false)
```

Comment:
If we are on a update, we need to call something similar to the line 248 above, passing both `new` and `current`, to ensure the new fields will be merged on top of current. Maybe changing [this line](https://github.com/eniolopes/ecto/blob/469f1f90f1647e70c823741c572f10e869ac1faa/lib/ecto/changeset/relation.ex#L247) to `if new_pks.(new) == current_pks.(current) or relation.on_change == :update do` is a better way to move forward?